### ISSUE TITLE
Named and nameless types.

### ISSUE THREAD

**Issue by Ericson2314** [haskell/cabal#5258]
We have a general principle in software engineering that things shouldn't know there own names. In the the DB case, that means talking about rows without their "surrogate key" with trick's like Persistent's `Entity`.

I think we could benefit from something similar in Cabal. Take for example, components, under the assumption I've finally finished my https://github.com/haskell/cabal/pull/5100 separating GDP from PD :D. We replace `Library` with `type NamedLibrary = (UnqualComponent, Library)` and so on. A GDP could have quite invalid structure, so it should stay the same using `NamedLibrary` where `Library` used before, but a PD should ensure that names are not reassigned to components and instead have a `Map UnqualComponent Library` with a convenience for `NamedLibrary`.

This and similar tricks would allow us to enforce a bunch of invariants that should be made true in transformations like GDP -> PD.

**Comment by 23Skidoo** [comment]
The idea sounds sensible to me.

**Comment by Ericson2314** [comment]
Relatedly, I'd like to parse with `[(UnqualComponent, CondTree Libray)]`, and then convert to  `Map UnqualComponent (CondTree Libray)`. This mirrors what we could do with my unmerged `LibDependency` and @fgaz's `Dependency` in that that is also grouping by package name.

**Comment by phadej** [comment]
@Ericson2314 what's status with this. Is this post-3.0 now?

**Comment by Ericson2314** [comment]
@phadej Yeah I haven't started on it. Should I try to bang it out quick? :D

**Comment by 23Skidoo** [comment]
Probably no

**Comment by Ericson2314** [comment]
That's what I figured. My only concern for this and a few other things is what the situation wrt breaking changes will be after 3.

**Comment by 23Skidoo** [comment]
We can have breaking changes in the next major release (3.2).

**Comment by Ericson2314** [comment]
OK great!

**Comment by ulysses4ever** [comment]
This looks like a discussion that concluded. Should we close this issue?
